src/main/java/org/gaul/s3proxy/ShardedBlobStore.java

```diff
@@ -0,0 +1,677 @@
+                        bucket));
+            shardsMap.put(bucket, shards);
+        }
+        return shardsMap.build();
+    }
+
+    public static ImmutableMap<String, String> parsePrefixes(
+            Properties properties) {
+        ImmutableMap.Builder<String, String> prefixesMap =
+                new ImmutableMap.Builder<>();
+        for (String key : properties.stringPropertyNames()) {
+            Matcher matcher = PROPERTIES_PREFIX_RE.matcher(key);
+            if (!matcher.matches()) {
+                continue;
+            }
+            prefixesMap.put(matcher.group("bucket"),
+                    properties.getProperty(key));
+        }
+        return prefixesMap.build();
+    }
+
+    static ShardedBlobStore newShardedBlobStore(
+            BlobStore blobStore,
+            ImmutableMap<String, Integer> shards,
+            ImmutableMap<String, String> prefixes) {
+        return new ShardedBlobStore(blobStore, shards, prefixes);
+    }
+
+    private Map<String, String> createSuperblockMeta(ShardedBucket bucket) {
+        ImmutableMap.Builder<String, String> meta =
+                new ImmutableMap.Builder<>();
+        meta.put("s3proxy-sharded-superblock-version", SUPERBLOCK_VERSION);
+        meta.put("s3proxy-sharded-superblock-prefix", bucket.prefix);
+        meta.put("s3proxy-sharded-superblock-shards",
+                Integer.toString(bucket.shards));
+        return meta.build();
+    }
+
+    private static String getShardContainer(ShardedBucket bucket, int shard) {
+        return String.format("%s-%d", bucket.prefix, shard);
+    }
+
+    private String getShard(String containerName, String blob) {
+        ShardedBucket bucket = buckets.get(containerName);
+        if (bucket == null) {
+            return containerName;
+        }
+        HashCode hash = SHARD_HASH.hashString(blob, StandardCharsets.UTF_8);
+        return ShardedBlobStore.getShardContainer(
+                bucket, Hashing.consistentHash(hash, bucket.shards));
+    }
+
+    private void checkSuperBlock(Blob blob, Map<String, String> expectedMeta,
+                                 String container) {
+        Map<String, String> currentSuperblockMeta =
+                blob.getMetadata().getUserMetadata();
+        for (String key : expectedMeta.keySet()) {
+            String current = currentSuperblockMeta.get(key);
+            String expected = expectedMeta.get(key);
+            if (!expected.equalsIgnoreCase(current)) {
+                throw new RuntimeException(String.format(
+                        "Superblock block for %s does not match: %s, %s",
+                        container, expected, current));
+            }
+        }
+    }
+
+    private boolean createShards(ShardedBucket bucket, Location location,
+                                 CreateContainerOptions options) {
+        Exception err = null;
+        ImmutableMap.Builder<Future<Boolean>, String> futuresMapBuilder =
```

Comment:
This seems backwards to me -- did you intend to map from container name to the `Future`?

Reply:
The intent here was correct -- I needed a way to map a future back to the corresponding shard. However, I removed the logging of individual failures, which obviated the need for this structure entirely. This Map is replaced with a List in the new revision.